src/Interpreters/Streaming/AggregateDataEx.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <Common/serde.h>
+
+namespace DB
+{
+using AggregateDataPtr = char *;
+using ConstAggregateDataPtr = const char *;
+
+namespace Streaming
+{
+SERDE struct UpdatedDataEx
+{
+    static ALWAYS_INLINE UpdatedDataEx & data(AggregateDataPtr __restrict place) { return *reinterpret_cast<UpdatedDataEx *>(place); }
+    static ALWAYS_INLINE const UpdatedDataEx & data(ConstAggregateDataPtr __restrict place) { return *reinterpret_cast<const UpdatedDataEx *>(place); }
+
+    static ALWAYS_INLINE bool isEmpty(ConstAggregateDataPtr __restrict place) { return data(place).final_count == 0; }
+    static ALWAYS_INLINE bool isUpdated(ConstAggregateDataPtr __restrict place) { return data(place).updated_since_last_finalization; }
+    static ALWAYS_INLINE void setUpdated(AggregateDataPtr __restrict place) { data(place).updated_since_last_finalization = true; }
+    static ALWAYS_INLINE void resetUpdated(AggregateDataPtr __restrict place) { data(place).updated_since_last_finalization = false; }
+
+    static void addBatch(size_t row_begin, size_t row_end, AggregateDataPtr * places, const IColumn * delta_col)
+    {
+        if (delta_col == nullptr)
```

Comment:
If we are tracking changes already, is `delta_col` always there ?

Reply:
nope, only exists for changelog input. but for `no changelog input` and `emit changelog`, we need tracking updates with retract, the `delta_col` is null

Reply:
For `TrackingUpdates`, there is always delta_col.
For `TrackingUpdatesWithRetract`, delta_col may be nullptr.
But I don’t think it’s worth implementing two codes